std/experimental/allocator/building_blocks/ascending_page_allocator.d

```diff
@@ -0,0 +1,418 @@
+module std.experimental.allocator.building_blocks.ascending_page_allocator;
+import std.experimental.allocator.common;
+/**
+`AscendingPageAllocator` is a fast and safe allocator which rounds all allocations
+to page size multiples. It reserves a range of virtual addresses (`mmap` for Posix
+and `VirtualAlloc` for Windows) and allocates memory at consecutive virtual
+addresses.
+
+When a chunk of memory is requested, the allocator finds a range of
+virtual pages which satisfy the requested size, changing their protection to
+read and write using OS primitives (`mprotect` and `VirtualProtect`).
+The physical memory is allocated on demand, when the pages are accessed.
+
+Deallocation removes any read/write permissions from the target pages
+and notifies the OS to reclaim the physical memory, while keeping the virtual
+memory.
+*/
+struct AscendingPageAllocator
+{
+    enum size_t pageSize = 4096;
+    size_t numPages;
+    bool valid;
+
+    // The start of the virtual address range
+    void* data;
+
+    // Keeps track of there the next allocation should start
+    void* offset;
+
+    // Number of pages which contain alive objects
+    size_t pagesUsed;
+
+    /**
+    The allocator receives as a parameter the size in pages of the virtual
+    address range (we assume the page size is 4096).
+    */
+    this(size_t pages)
+    {
+        import std.exception : enforce;
+
+        valid = true;
+        numPages = pages;
+        version(Posix)
+        {
+            import core.sys.posix.sys.mman : mmap, MAP_ANON, PROT_READ,
+                   PROT_WRITE, PROT_NONE, MAP_PRIVATE, MAP_FAILED;
+            data = mmap(null, pageSize * pages, PROT_NONE, MAP_PRIVATE | MAP_ANON, -1, 0);
+            enforce(data != MAP_FAILED, "Failed to mmap memory");
+        }
+        else version(Windows)
+        {
+            import core.sys.windows.windows : VirtualAlloc, PAGE_NOACCESS,
+                   MEM_COMMIT, MEM_RESERVE;
+            data = VirtualAlloc(null, pageSize * pages, MEM_COMMIT | MEM_RESERVE, PAGE_NOACCESS);
+            enforce(data != null, "Failed to VirtualAlloc memory");
+        }
+        offset = data;
+    }
+
+    /**
+    Round the allocation to the next multiple of the page size.
+    The allocation only reserves a range of virtual pages but the actual
+    physical memory is allocated on demand, when accessing the memory
+    Returns `null` on failure or if the requested size exceeds the remaining capacity.
+    */
+    void[] allocate(size_t n)
+    {
+        import std.exception : enforce;
+
+        size_t goodSize = goodAllocSize(n);
+        if (offset - data > numPages * pageSize - goodSize)
```

Comment:
pedantically nicer because it has fewer chance of overflowing: `if (offset + goodSize > data + numPages * pageSize)`